tidy3d/components/source.py

```diff
@@ -291,7 +294,83 @@ def amp_time(self, time: float) -> complex:
         return const * offset * oscillation * amp
 
 
-SourceTimeType = Union[GaussianPulse, ContinuousWave]
+class CustomSourceTime(Pulse):
+    """Custom source time dependence.
+
+    Note
+    ----
+    To interpolate to simulation time steps, the value provided at the nearest time is used.
+    In order to ensure that the source time dependence is smooth, it is recommended to either
+    provide source time data exactly at the simulation time steps, or alternatively to oversample
+    by providing source time data at a much closer spacing than the simulation time step.
+
+    Example
+    -------
+    >>> cst = td.CustomSourceTime.from_values(freq0=1, fwidth=0.1,
+    >>>     values=np.linspace(0, 9, 10), dt=0.1)
+    """
+
+    source_time_dataset: Optional[TimeDataset] = pydantic.Field(
+        ..., title="Source time dataset", description="Dataset for storing the custom source time."
+    )
+
+    _source_time_dataset_none_warning = warn_if_dataset_none("source_time_dataset")
+
+    @classmethod
+    def from_values(
+        cls, freq0: float, fwidth: float, values: ArrayComplex1D, dt: float
+    ) -> CustomSourceTime:
+        """Create a CustomSourceTime from a numpy array.
+
+        Parameters
+        ----------
+        freq0 : float
+            Estimated central frequency of the source.
+        fwidth : float
+            Estimated frequency width of the source.
+        values: ArrayComplex1D
+            Complex values of the source amplitude.
+        dt: float
+            Time step for the `values` array. It is recommended to either use the simulation `dt`
+            or a value much smaller than the simulation `dt` in order to ensure the source time
+            dependence is smooth.
+
+
+        Returns
+        -------
+        CustomSourceTime
+            :class:`.CustomSourceTime` with these values, and time coordinates evenly spaced
+            between 0 and dt * (N-1) with a step size of `dt`, where N is the length of
+            the values array.
+        """
+        times = np.arange(len(values)) * dt
+        source_time_dataarray = TimeDataArray(values, coords=dict(t=times))
+        source_time_dataset = TimeDataset(values=source_time_dataarray)
+        return CustomSourceTime(
+            freq0=freq0,
+            fwidth=fwidth,
+            source_time_dataset=source_time_dataset,
+        )
+
+    def amp_time(self, time: float) -> complex:
+        """Complex-valued source amplitude as a function of time.
+
+        Parameters
+        ----------
+        time : float
+            Time in seconds.
+
+        Returns
+        -------
+        complex
+            Complex-valued source amplitude at that time.
+        """
+        if self.source_time_dataset is None:
+            return None
+        return self.source_time_dataset.values.sel(t=time, method="nearest").to_numpy()
+
+
+SourceTimeType = Union[GaussianPulse, ContinuousWave, CustomSourceTime]
```

Comment:
note: maybe we should just take this opportunity to remove CW,  dont think it's actually supported. @momchil-flex ?

Reply:
CW works, it just isn't shown in the docs

Reply:
Yeah it should work, it's not tested though.